### ISSUE TITLE
Building android project hangs forever

### ISSUE THREAD

**Issue by stesvis** [xamarin/xamarin-android#5936]
"Suddenly" i am unable to build **android** projects in VS on **windows**.
I think it started after some latest Windows 10 updates, but I tried to roll back and uninstall them and I wasn't able to build even in that case.
I've seen this same issue in a ticket in the past, however those solutions didn't work and comments were closed, so my option now is to open a new issue...

### Steps to Reproduce

1. Get latest VS version (16.9.6) and latest Windows 10 updates
2. Create a blank xamarin project (i created it as xamarin forms)
3. Build android

### Expected Behavior
Project builds

### Actual Behavior
Projects hangs and never finishes to build. I can reproduce it with all my existing apps that were able to build last week, and with brand new empty apps.

### Version Information

Microsoft Visual Studio Enterprise 2019 (2)
Version 16.9.6
VisualStudio.16.Release/16.9.6+31313.79
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Enterprise

.NET Core Debugging with WSL 2   1.0
.NET Core Debugging with WSL 2

ASP.NET and Web Tools 2019   16.9.693.2781
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2112521+5741df381174d72f08e3632bb99f52e8635b6a1a
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.9.693.2781
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.9.693.2781
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.9.693.2781
Azure Functions and Web Jobs Tools

C# Tools   3.9.0-6.21160.10+59eedc33d35754759994155ea2f4e1012a9951e3
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

CodeMaid   11.2.231
CodeMaid is an open source Visual Studio extension to cleanup and simplify our C#, C++, F#, VB, PHP, PowerShell, R, JSON, XAML, XML, ASP, HTML, CSS, LESS, SCSS, JavaScript and TypeScript coding.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.40423.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.113+g422d40002e.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Mono Debugging for Visual Studio   16.9.7 (df23ba6)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.9.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Productivity Power Tools 2017/2019   16.0
Installs the individual extensions of Productivity Power Tools 2017/2019

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Snapshot Debugging Extension   1.0
Snapshot Debugging Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.62103.10080
Microsoft SQL Server Data Tools

TGit   1.0
Control TortoiseGit from within Visual Studio

TypeScript Tools   16.0.30201.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.9.0-6.21160.10+59eedc33d35754759994155ea2f4e1012a9951e3
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   16.9.0-beta.21102.9+7ce7132f1459095e635194d09d6f73265352029a
Microsoft Visual F# Tools

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Foo   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VSColorOutput   2.73
Color output for build and debug windows - https://mike-ward.net/vscoloroutput

Xamarin   16.9.000.294 (d16-9@876547a)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.9.0.318 (remotes/origin/7b35adcdd1b204bbdcb279e0f029d151a20a4bf3@7b35adcdd)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.9.72 (426ebf6)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   11.2.2.1 (d16-9/877f572)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: 5e9cb6d
    Java.Interop: xamarin/java.interop/d16-9@54f8c24
    ProGuard: Guardsquare/proguard/v7.0.1@912d149
    SQLite: xamarin/sqlite/3.34.1@daff8f4
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-9@d210f11

Xamarin.iOS and Xamarin.Mac SDK   14.16.0.5 (ab40b131d)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

### Log File

Unfortunately, the log file is not generated, when i run that command, again, it hangs forever:
![image](https://user-images.githubusercontent.com/6062684/118705395-69731980-b7d5-11eb-9e2b-0b42b0aa62f9.png)

**Comment by dellis1972** [comment]
@stesvis 

So from the Adb Command Prompt if you just go `adb logcat -d` without the `clip` it still hangs?

It would also be useful to get a Diagnostic build log which would help us see where in the build/deploy process it is hanging.
See https://docs.microsoft.com/en-gb/xamarin/android/troubleshooting/troubleshooting#diagnostic-msbuild-output for details on how to get that data. Note that in the Output Window there will be options for both Build and Deploy, we will need both.

**Comment by dellis1972** [comment]
Also just for testing it might be worth seeing if you can turn off the Virus Scanner you have running to see if that might be making a difference? or testing with and without the firewall on. Its possible something is blocking network traffic and `adb` uses ports for all of its communications.

Also if you can find the `adb.exe` file on your system, right click on it and select `Properties`, then check to see if there is an `Unblock` check box near the bottom of the property page. It might be that windows now thinks it is a dangerous exe or something.

**Comment by stesvis** [comment]
> @stesvis
> 
> So from the Adb Command Prompt if you just go `adb logcat -d` without the `clip` it still hangs?
> 
> It would also be useful to get a Diagnostic build log which would help us see where in the build/deploy process it is hanging.
> See https://docs.microsoft.com/en-gb/xamarin/android/troubleshooting/troubleshooting#diagnostic-msbuild-output for details on how to get that data. Note that in the Output Window there will be options for both Build and Deploy, we will need both.

Hi @dellis1972  yes it still hangs without the `clip` option.
Here's the log: https://gist.github.com/stesvis/ae19de9365d3b511aa996ebf1d35357e

In the output window I only have option to `Build`, not `Deploy`:

![image](https://user-images.githubusercontent.com/6062684/118843301-27a2ab80-b887-11eb-90f8-1c61e5e7d2d8.png)

**Comment by stesvis** [comment]
> Also just for testing it might be worth seeing if you can turn off the Virus Scanner you have running to see if that might be making a difference? or testing with and without the firewall on. Its possible something is blocking network traffic and `adb` uses ports for all of its communications.
> 
> Also if you can find the `adb.exe` file on your system, right click on it and select `Properties`, then check to see if there is an `Unblock` check box near the bottom of the property page. It might be that windows now thinks it is a dangerous exe or something.

Same, I disabled firewall and Windows defender and it still hangs

**Comment by dellis1972** [comment]
Next thing to do is to see if you have any processes using port 5037 running. Run `netstat -aon`  in an Administrator command prompt. That should give you a PID which you can use to figure out which process it is.

**Comment by stesvis** [comment]
These are the references to port 5037:

```
  TCP    127.0.0.1:5037         127.0.0.1:50900        ESTABLISHED     4712
  TCP    127.0.0.1:5037         127.0.0.1:51760        ESTABLISHED     4712
  TCP    127.0.0.1:5939         0.0.0.0:0              LISTENING       4872
  TCP    127.0.0.1:27015        0.0.0.0:0              LISTENING       13776
  TCP    127.0.0.1:27015        127.0.0.1:50895        ESTABLISHED     13776
  TCP    127.0.0.1:50889        127.0.0.1:51647        ESTABLISHED     4892
  TCP    127.0.0.1:50891        127.0.0.1:51647        ESTABLISHED     10868
  TCP    127.0.0.1:50893        127.0.0.1:51647        ESTABLISHED     13888
  TCP    127.0.0.1:50895        127.0.0.1:27015        ESTABLISHED     13888
  TCP    127.0.0.1:50900        127.0.0.1:5037         ESTABLISHED     10868
  TCP    127.0.0.1:51614        127.0.0.1:1025         ESTABLISHED     17320
  TCP    127.0.0.1:51647        127.0.0.1:50889        ESTABLISHED     4892
  TCP    127.0.0.1:51647        127.0.0.1:50891        ESTABLISHED     4892
  TCP    127.0.0.1:51647        127.0.0.1:50893        ESTABLISHED     4892
  TCP    127.0.0.1:51760        127.0.0.1:5037         ESTABLISHED     11216
  TCP    [::]:5037                     [::]:0                         LISTENING       4712
 ```

**Comment by dellis1972** [comment]
@stesvis so you need to figure out which processes are responsible for those. I think you can see the Process Id in Task Manager for each process.

**Comment by stesvis** [comment]
@dellis1972 sure i can find it but i am not sure what you want me to do, kill that process and try to build?
The process with PID 4712 is `java.exe`

**Comment by dellis1972** [comment]
I would say yes try killing the process. Not sure why java would be holding that port.

**Comment by stesvis** [comment]
After killing `java.exe` when I rebuild the project, VS freezes.
![image](https://user-images.githubusercontent.com/6062684/118853135-a9e39d80-b890-11eb-8ec1-b6992712f824.png)

**Comment by dellis1972** [comment]
ok. Next thing to try is adding a System Environment Variable for `ANDROID_ADB_SERVER_PORT`. Select a different value to 5037 say 5038. You will probably need to close VS and any Developer Command Prompts first, then set the System Environment variable then start VS again. 

I've been doing some research and it seems most of the problems with `adb` hanging are to do with port clashes.

**Comment by stesvis** [comment]
Thanks @dellis1972 , but unfortunately it's still hanging.

**Comment by dellis1972** [comment]
@stesvis  hmm this is really really weird. 

Lets take a step back and rule somethings out. 
First close VS. Then kill adb via `adb kill-server` on the command line (Assuming adb is in your path).


Then from the Developer command prompt can you run this

```
msbuild YouRent.Mobile.Android.csproj /restore /t:SignAndroidPackage /v:diag /p:Configuration=Release
```

I'm trying to figure out if it is just adb which is hanging or other android tooling.

**Comment by stesvis** [comment]
@dellis1972 thanks for the support on this:
`adb kill-server` gives me this error `cannot connect to daemon at tcp:5037: cannot connect to 127.0.0.1:5037: No connection could be made because the target machine actively refused it. (10061)` (or tcp:5038 if i set that port number).

Anyway, i ran the other command and it completed successfully. After that, I built the project using the IDE build tools as usual and surprise surprise, it worked this time (in `Release` and in `Debug` mode)!

Hopefully it will keep working even after the next restart, but at least for now i am back to business, so THANK YOU!
Can you explain what that command does?

**Comment by dellis1972** [comment]
@stesvis 

The way `adb` works is when you call it , it will launch a background process which will look out for devices etc and process commands. When you then make calls to `adb` manually (or via our build process) it will talk to the background process over tcp.
So the `adb kill-server` is supposed to force the shutdown of the `adb` background process. 

I assume is was the `kill-server` command you wanted an explanation for? Hope it helps.
I'm going to close this Issue for now as it seems to be working, please re-open it if you hit problems again.

**Comment by stesvis** [comment]
@dellis1972 Unfortunately it's happening again and those commands from last time are not helpng.
`adb kill-server` hangs forever so i have to cancel it with `Ctrl+C`
`msbuild` is hanging too and this is where:

> Using "GetPrimaryCpuAbi" task from assembly "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Xam
> arin\Android\Xamarin.Android.Build.Debugging.Tasks.dll".
> Task "GetPrimaryCpuAbi" (TaskId:78)
>   Task Parameter:AndroidPackage=ca.levitica.livedispatch.prod (TaskId:78)
>   Task Parameter:DevicePropertyCache=obj\Release\90\devices.cache (TaskId:78)

I am really stuck and I can't even make an APK for the store...

**Comment by dellis1972** [comment]
I'm stumped as to why you are having this issue. 
The only thing I can think of is to download Process Explorer from https://docs.microsoft.com/en-gb/sysinternals/downloads/process-explorer. This will allow you to check what processes are running (and kill them) , I think it should also allow you to see what ports are being used and by what.

I would suggest shutting down VS. Then using process Explorer to kill off any adb.exe which is left running. Then try again to build your app from the command line as we did before.

**Comment by dellis1972** [comment]
These is also PortMon https://docs.microsoft.com/en-gb/sysinternals/downloads/portmon for looking at networking stuff.

**Comment by stesvis** [comment]
Ok thanks @dellis1972 so the goal is to kill `adb.exe` and then retry?
If that is the case, i don't have any `adb.exe` process running at the moment and it still fails at `Task "GetPrimaryCpuAbi" (TaskId:80)`

**Comment by stesvis** [comment]
After a bunch of tests, it was stuck at that point i said in the previous comment, and when I killed `java.exe` it resumed on it's own and completed the build... I am like... `???`

**Comment by dellis1972** [comment]
> After a bunch of tests, it was stuck at that point i said in the previous comment, and when I killed `java.exe` it resumed on it's own and completed the build... I am like... `???`

Errr ... WT actual F. Why is java stopping adb working, that's nuts.

**Comment by dellis1972** [comment]
@stesvis I'm wondering if something on your machine is running java and its keeping the 5037 port locked. 
When you are trying to build, what is running other than Visual Studio? 
If you download Process Explorer I mentioned earlier you can probably also see what arguments the java.exe was started with. That way we might be able to track down what is starting that process.

**Comment by stesvis** [comment]
@dellis1972 when I run Visual Studio I usually have MS Outlook, MS Teams, sometimes SSMS, VS Code, Postman, Slack....
But I always had these programs open also in the past.

I did install Process Explorer, it says "parent: devenv.exe"....:

![image](https://user-images.githubusercontent.com/6062684/120664836-7e43e400-c448-11eb-9abe-53ddcd9e9633.png)

**Comment by dellis1972** [comment]
@stesvis ok, so its working for now? What I would suggest is when it does hang again take a look in Process Explorer and see what java is doing. Its odd VS would kick off java and leave it hanging around, unless there is a java project loaded somewhere. 
All of our build system which does use java we fire it up and close it after use.

**Comment by stesvis** [comment]
@dellis1972 it definitely looks like every time i restart the PC i have to kill `java.exe` and then it works again. I have no other way, and it all started a couple of weeks ago. Before i never had an issue.
